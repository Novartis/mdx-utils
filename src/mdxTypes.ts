export enum MdxNodeType {
  Root = 'root',
  Heading = 'heading',
  Paragraph = 'paragraph',
  ThematicBreak = 'thematicBreak',
  CodeBlock = 'code',
  List = 'list',
  ListItem = 'listItem',
  Blockquote = 'blockquote',
  Yaml = 'yaml',
  Link = 'link',
  Strong = 'strong',
  Em = 'emphasis',
  InlineCode = 'inlineCode',
  Import = 'import',
  Export = 'export',
  Jsx = 'jsx',
  Text = 'text',
}

export type RootNode = {
  type: MdxNodeType.Root;
  children: Array<MdxNode>;
  meta?: YamlValue;
};
export type TextNode = { type: MdxNodeType.Text; value: string };
export type HeadingNode = {
  type: MdxNodeType.Heading;
  depth: number;
  children: Array<MdxNode>;
};
export type ParagraphNode = {
  type: MdxNodeType.Paragraph;
  children: Array<MdxNode>;
};
export type BlockquoteNode = {
  type: MdxNodeType.Blockquote;
  children: Array<MdxNode>;
};
export type ListNode = {
  type: MdxNodeType.List;
  ordered: boolean;
  start: number | null;
  spread: boolean;
  children: Array<MdxNode>;
};
export type ListItemNode = {
  type: MdxNodeType.ListItem;
  /**
   * Does this list item have empty lines?
   *
   * Previously known as "loose" and is easier to Google that way.
   */
  spread: boolean;
  checked: null | boolean;
  children: Array<MdxNode>;
};
export type CodeBlockNode = {
  type: MdxNodeType.CodeBlock;
  lang: string;
  meta: string | null;
  value: string;
};
export type ThematicBreakNode = { type: MdxNodeType.ThematicBreak };
export type YamlNode = {
  type: MdxNodeType.Yaml;
  value: string;
};
export type LinkNode = {
  type: MdxNodeType.Link;
  title: null | string;
  url: string;
  children: Array<MdxNode>;
};
export type StrongNode = {
  type: MdxNodeType.Strong;
  children: Array<MdxNode>;
};
export type EmNode = { type: MdxNodeType.Em; children: Array<MdxNode> };
export type InlineCodeNode = {
  type: MdxNodeType.InlineCode;
  value: string;
};

export type ImportNode = { type: MdxNodeType.Import; value: string };
export type ExportNode = { type: MdxNodeType.Export; value: string };
export type JsxNode = { type: MdxNodeType.Jsx; value: string };

export type MdxNode =
  | RootNode
  | TextNode
  | YamlNode
  | HeadingNode
  | ParagraphNode
  | ListNode
  | ListItemNode
  | ThematicBreakNode
  | BlockquoteNode
  | CodeBlockNode
  | LinkNode
  | StrongNode
  | EmNode
  | InlineCodeNode
  | ImportNode
  | ExportNode
  | JsxNode;

export type YamlValue<T = never> =
  | string
  | number
  | boolean
  | T
  | { [key: string]: YamlValue<T> }
  | Array<YamlValue<T>>;
