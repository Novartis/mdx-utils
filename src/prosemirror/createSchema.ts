import { Schema, NodeSpec, MarkSpec } from 'prosemirror-model';
import { MarkdownNodeType, MarkdownMarkType } from 'prosemirror-markdown';
import { createMarksConfig } from './createMarksConfig';
import { createNodesConfig } from './createNodesConfig';

export default function createSchema<
  CustomNodes extends string,
  CustomMarks extends string
>({
  nodes,
  marks,
}: {
  nodes: Record<CustomNodes, NodeSpec>;
  marks: Record<CustomMarks, MarkSpec>;
}): Schema<CustomNodes | MarkdownNodeType, CustomMarks | MarkdownMarkType> {
  return new Schema<
    CustomNodes | MarkdownNodeType,
    CustomMarks | MarkdownMarkType
  >({
    nodes: createNodesConfig(nodes),
    marks: createMarksConfig(marks),
  });
}
