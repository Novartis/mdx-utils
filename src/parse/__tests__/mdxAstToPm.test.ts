import createMdxConverter from '../mdxAstToPm';
import createSchema from '../../prosemirror/createSchema';
import { Node } from 'prosemirror-model';
import { MdxNodeType, JsxNode } from '../../mdxTypes';
import parseMdxToAst from '../parseMdx';

const schema = createSchema({
  nodes: {
    my_custom_component: {
      atom: true,
      attrs: {
        count: {},
      },
      toDOM() {
        return ['div', { class: 'my_custom_component' }];
      },
    },
  },
  marks: {},
});
const mdxAstToPm = createMdxConverter(schema, { jsxToNode });
function parseMdx(source: string) {
  const ast = parseMdxToAst(source);
  return mdxAstToPm(ast);
}

/** Custom callback that turns JSX into custom nodes. */
function jsxToNode(node: JsxNode): null | Node<typeof schema> {
  if (node.value.startsWith('<MyCustomComponent')) {
    // In a real implementation, you would use Babel or another real parser to extract information.
    return schema.nodes.my_custom_component.create({ count: 1 });
  }
  return null;
}

describe('mdxAstToPm', () => {
  it('should parse an example document with YAML', () => {
    const pmTree = parseMdx(`---
title: Design System
---

# Design System

> Design system for NIBR`);
    expect(pmTree.attrs.frontmatter).toEqual({ title: 'Design System' });

    pmTree.forEach((node) => {
      // Triple-hyphen should parse as YAML delimeter, not <hr>
      expect(node.type.name === MdxNodeType.ThematicBreak).toBeFalsy();
    });
  });

  it('should parse block JSX nodes', () => {
    const ast = parseMdx(`---
title: Example JSX
---
import Button from '@material-ui/core/Button';

<Button color="primary">Click me</Button>

`);
    const children = Array.from({ length: ast.childCount }, (_, i) =>
      ast.child(i)
    );
    expect(children.find(({ type }) => type.name === 'mdx_import').toJSON())
      .toMatchInlineSnapshot(`
      Object {
        "attrs": Object {
          "value": "import Button from '@material-ui/core/Button';",
        },
        "type": "mdx_import",
      }
    `);
    expect(children.find(({ type }) => type.name === 'mdx_jsx').toJSON())
      .toMatchInlineSnapshot(`
      Object {
        "attrs": Object {
          "value": "<Button color=\\"primary\\">Click me</Button>",
        },
        "type": "mdx_jsx",
      }
    `);
  });

  it('should use custom parser for provided JSX nodes', () => {
    const ast = parseMdx(`---
title: Custom components
---

<MyCustomComponent prop="value" />

`);
    const children = Array.from({ length: ast.childCount }, (_, i) =>
      ast.child(i)
    );
    expect(children.find((node) => node.type.name === 'my_custom_component'))
      .toMatchInlineSnapshot(`
      Object {
        "attrs": Object {
          "count": 1,
        },
        "type": "my_custom_component",
      }
    `);
  });

  it('should parse most node types', () => {
    const ast = parseMdx(`---
title: MDX nodes
---

import CustomComponent from './CustomComponent';

# MDX nodes

Here is an example of many different MDX elements. You can have **bold** text, _italic_ text, \`code\`, and [links](https://novartis.com).

Here's a list:

- First item
- Second item
- Third item
    1. This one has a nested ordered list
    2. Should parse correctly

> Sometimes you like to use blockquotes as well

----

## More examples

<CustomComponent prop="value" />

export function hello() {
  return 'world';
}

\`\`\`js
// Code sample
\`\`\`

`);
    expect(ast).toMatchSnapshot();
  });
});
