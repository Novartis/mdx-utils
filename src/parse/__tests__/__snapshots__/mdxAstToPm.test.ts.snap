// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mdxAstToPm should parse most node types 1`] = `
Object {
  "attrs": Object {
    "frontmatter": Object {
      "title": "MDX nodes",
    },
  },
  "content": Array [
    Object {
      "attrs": Object {
        "value": "import CustomComponent from './CustomComponent';",
      },
      "type": "mdx_import",
    },
    Object {
      "attrs": Object {
        "level": 1,
      },
      "content": Array [
        Object {
          "text": "MDX nodes",
          "type": "text",
        },
      ],
      "type": "heading",
    },
    Object {
      "content": Array [
        Object {
          "text": "Here is an example of many different MDX elements. You can have ",
          "type": "text",
        },
        Object {
          "marks": Array [
            Object {
              "type": "strong",
            },
          ],
          "text": "bold",
          "type": "text",
        },
        Object {
          "text": " text, ",
          "type": "text",
        },
        Object {
          "marks": Array [
            Object {
              "type": "em",
            },
          ],
          "text": "italic",
          "type": "text",
        },
        Object {
          "text": " text, ",
          "type": "text",
        },
        Object {
          "marks": Array [
            Object {
              "type": "code",
            },
          ],
          "text": "code",
          "type": "text",
        },
        Object {
          "text": ", and ",
          "type": "text",
        },
        Object {
          "marks": Array [
            Object {
              "attrs": Object {
                "href": "https://novartis.com",
                "title": null,
              },
              "type": "link",
            },
          ],
          "text": "links",
          "type": "text",
        },
        Object {
          "text": ".",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "text": "Here's a list:",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "attrs": Object {
        "tight": true,
      },
      "content": Array [
        Object {
          "content": Array [
            Object {
              "content": Array [
                Object {
                  "text": "First item",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "list_item",
        },
        Object {
          "content": Array [
            Object {
              "content": Array [
                Object {
                  "text": "Second item",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "list_item",
        },
        Object {
          "content": Array [
            Object {
              "content": Array [
                Object {
                  "text": "Third item",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
            Object {
              "attrs": Object {
                "order": 1,
                "tight": true,
              },
              "content": Array [
                Object {
                  "content": Array [
                    Object {
                      "content": Array [
                        Object {
                          "text": "This one has a nested ordered list",
                          "type": "text",
                        },
                      ],
                      "type": "paragraph",
                    },
                  ],
                  "type": "list_item",
                },
                Object {
                  "content": Array [
                    Object {
                      "content": Array [
                        Object {
                          "text": "Should parse correctly",
                          "type": "text",
                        },
                      ],
                      "type": "paragraph",
                    },
                  ],
                  "type": "list_item",
                },
              ],
              "type": "ordered_list",
            },
          ],
          "type": "list_item",
        },
      ],
      "type": "bullet_list",
    },
    Object {
      "content": Array [
        Object {
          "content": Array [
            Object {
              "text": "Sometimes you like to use blockquotes as well",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
      "type": "blockquote",
    },
    Object {
      "type": "horizontal_rule",
    },
    Object {
      "attrs": Object {
        "level": 2,
      },
      "content": Array [
        Object {
          "text": "More examples",
          "type": "text",
        },
      ],
      "type": "heading",
    },
    Object {
      "attrs": Object {
        "value": "<CustomComponent prop=\\"value\\" />",
      },
      "type": "mdx_jsx",
    },
    Object {
      "attrs": Object {
        "value": "export function hello() {
  return 'world';
}",
      },
      "type": "mdx_export",
    },
    Object {
      "attrs": Object {
        "params": "",
      },
      "content": Array [
        Object {
          "text": "// Code sample",
          "type": "text",
        },
      ],
      "type": "code_block",
    },
  ],
  "type": "doc",
}
`;
